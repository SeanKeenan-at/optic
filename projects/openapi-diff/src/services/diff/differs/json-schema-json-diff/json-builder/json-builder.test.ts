import { opticJsonSchemaDiffer } from '../index';
import { streamingJsonSchemaBuilder } from './streaming-json-schema-builder';
import { locations } from '../plugins/tests/json-schema-diff-patch-fixture';
const schemaDiffer = opticJsonSchemaDiffer();

function fixture(input: any, ...additional: any[]) {
  const jsonSchemaForInput = streamingJsonSchemaBuilder(
    schemaDiffer,
    input,
    ...additional
  );

  const diffBetweenGeneratedAndInput = schemaDiffer.compare(
    jsonSchemaForInput,
    input,
    locations.inAResponse,
    '',
    { collapseToFirstInstanceOfArrayDiffs: true }
  );
  return {
    input,
    jsonSchemaForInput,
    diffBetweenGeneratedAndInput,
  };
}

describe('json builder', () => {
  describe('primitives', () => {
    it('can build JSON from a string', async () => {
      const result = fixture('string value');
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });
    it('can build JSON from a boolean', async () => {
      const result = fixture(true);
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });
    it('can build JSON from a number', async () => {
      const result = fixture(1544);
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });

    it('can build JSON from a null', async () => {
      const result = fixture(null);
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });
  });
  describe('objects', () => {
    it('can learn an object with primitive fields', () => {
      const result = fixture({
        hello: 'world',
        age: 145,
      });
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });
    it('can learn nested objects', () => {
      const result = fixture({
        hello: 'world',
        age: 145,
        nested: {
          nested2: {
            nested3: {
              stringField: 'abc',
            },
          },
        },
      });
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });

    it('can learn that an object with primitive fields is polymorphic', () => {
      const result = fixture(
        {
          hello: 'world',
          age: 145,
        },
        {
          // hello: "world", // commented to show it is missing
          age: 145,
        },
        {
          // hello: "world",
          age: 145,
          admin: true,
        },
        // the repeats stress the system and make sure the bail from infinite loop logic works
        {
          // hello: "world",
          age: 145,
          admin: true,
        },
        {
          // hello: "world",
          age: 145,
          admin: true,
        },
        {
          // hello: "world",
          age: 145,
        }
      );
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });
  });
  describe('arrays', () => {
    it('can learn an array of object items', () => {
      const result = fixture([{ field: 1 }, { field: 2 }, { field: 3 }]);
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });
    it('can learn an array of object items with polymorphism', () => {
      const result = fixture([
        { field: 1 },
        { field: 2 },
        { field: 3, field2: '' },
      ]);
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });
    it('can learn an array of primitive items', () => {
      const result = fixture(['a', 'b', 'c']);
      expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
      expect(result).toMatchSnapshot();
    });

    // it("can learn an array of empty items, then a string", () => {
    //   const result = fixture([], ["hello"]);
    //   expect(result.diffBetweenGeneratedAndInput).toHaveLength(0);
    //   expect(result).toMatchSnapshot();
    // });
  });
});
