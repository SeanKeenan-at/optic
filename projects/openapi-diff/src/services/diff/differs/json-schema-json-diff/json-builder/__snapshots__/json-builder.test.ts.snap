// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json builder arrays can learn an array of object items 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": Array [
    Object {
      "field": 1,
    },
    Object {
      "field": 2,
    },
    Object {
      "field": 3,
    },
  ],
  "jsonSchemaForInput": Object {
    "items": Object {
      "properties": Object {
        "field": Object {
          "type": "number",
        },
      },
      "required": Array [
        "field",
      ],
      "type": "object",
    },
    "type": "array",
  },
}
`;

exports[`json builder arrays can learn an array of object items with polymorphism 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": Array [
    Object {
      "field": 1,
    },
    Object {
      "field": 2,
    },
    Object {
      "field": 3,
      "field2": "",
    },
  ],
  "jsonSchemaForInput": Object {
    "items": Object {
      "properties": Object {
        "field": Object {
          "type": "number",
        },
        "field2": Object {
          "type": "string",
        },
      },
      "required": Array [
        "field",
      ],
      "type": "object",
    },
    "type": "array",
  },
}
`;

exports[`json builder arrays can learn an array of primitive items 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": Array [
    "a",
    "b",
    "c",
  ],
  "jsonSchemaForInput": Object {
    "items": Object {
      "type": "string",
    },
    "type": "array",
  },
}
`;

exports[`json builder objects can learn an object with primitive fields 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": Object {
    "age": 145,
    "hello": "world",
  },
  "jsonSchemaForInput": Object {
    "properties": Object {
      "age": Object {
        "type": "number",
      },
      "hello": Object {
        "type": "string",
      },
    },
    "required": Array [
      "hello",
      "age",
    ],
    "type": "object",
  },
}
`;

exports[`json builder objects can learn nested objects 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": Object {
    "age": 145,
    "hello": "world",
    "nested": Object {
      "nested2": Object {
        "nested3": Object {
          "stringField": "abc",
        },
      },
    },
  },
  "jsonSchemaForInput": Object {
    "properties": Object {
      "age": Object {
        "type": "number",
      },
      "hello": Object {
        "type": "string",
      },
      "nested": Object {
        "properties": Object {
          "nested2": Object {
            "properties": Object {
              "nested3": Object {
                "properties": Object {
                  "stringField": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "stringField",
                ],
                "type": "object",
              },
            },
            "required": Array [
              "nested3",
            ],
            "type": "object",
          },
        },
        "required": Array [
          "nested2",
        ],
        "type": "object",
      },
    },
    "required": Array [
      "hello",
      "age",
      "nested",
    ],
    "type": "object",
  },
}
`;

exports[`json builder objects can learn that an object with primitive fields is polymorphic 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": Object {
    "age": 145,
    "hello": "world",
  },
  "jsonSchemaForInput": Object {
    "properties": Object {
      "admin": Object {
        "type": "boolean",
      },
      "age": Object {
        "type": "number",
      },
      "hello": Object {
        "type": "string",
      },
    },
    "required": Array [
      "age",
    ],
    "type": "object",
  },
}
`;

exports[`json builder primitives can build JSON from a boolean 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": true,
  "jsonSchemaForInput": Object {
    "type": "boolean",
  },
}
`;

exports[`json builder primitives can build JSON from a null 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": null,
  "jsonSchemaForInput": Object {
    "type": "null",
  },
}
`;

exports[`json builder primitives can build JSON from a number 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": 1544,
  "jsonSchemaForInput": Object {
    "type": "number",
  },
}
`;

exports[`json builder primitives can build JSON from a string 1`] = `
Object {
  "diffBetweenGeneratedAndInput": Array [],
  "input": "string value",
  "jsonSchemaForInput": Object {
    "type": "string",
  },
}
`;
